### ISSUE TITLE
Performance in Feat section

### ISSUE THREAD

**Issue by johannes-z** [Heliomance/HeroForge-Anew#50]
This is probably well known. The feat section performs very poorly - probably because of the huge amount of feats. 
Is there a way around to fix this? The only thing that comes into my mind would be splitting the feat section into multiple sections (f.e. giving the complete series feats it's own section)...

**Comment by torrasque666** [comment]
Might I suggest something akin to YAPCG's set up? Take the current list of feats, keep that as its own page, and add a new page to pull from the old information. The drop down would only have the feats from the sources selected and that are qualified for (as much as they can be. things that require a specific spell can only be assumed as possible based on class level and casting stat for instance. Or a story-based prereq) with a separate drop down detailing said prereqs. It would be nice if it could also list things in the order they are gained. For instance, a Level 4 Fighter's feat selection page would look like this:

Level 1 Feat:
Fighter 1 Bonus:
Fighter 2 Bonus:
Level 3:
Fighter 4 Bonus:

Flaws may need to add a drop down for the level they are selected at, to be placed in the proper area. (Technically, they only have to be taken during creation, not at level 1. Its just that the books assume Level 1 starts are the default.)

This should alleviate the processing power needed to run the Feats page, since I imagine it has something to do with having to all those objects.

I'll put together a rough draft in the morning.

**Comment by Heliomance** [comment]
The big problem with the Feats page is the huge numbers of checkboxes, I think. I definitely want to do something about it, but I'm low on ideas. I like the idea of dropdowns corresponding to the feat slots you actually have, but I'm drawing a blank on how to make it properly populate bonus feats.

**Comment by torrasque666** [comment]
I have a rough draft drawn up, and it populates well, but due to using mostly array formulas its incredibly slow. Image(s) below:

![image](https://cloud.githubusercontent.com/assets/10926262/9487957/d9747be0-4b9b-11e5-93f7-f2e89e55b8f4.png)
![image](https://cloud.githubusercontent.com/assets/10926262/9487968/e7f83d1e-4b9b-11e5-8ab5-58f90403ec22.png)
![image](https://cloud.githubusercontent.com/assets/10926262/9487976/f74a15da-4b9b-11e5-9d66-5dc47218145e.png)

**Comment by chaoscommencer** [comment]
If you guys are really concerned about performance, we could always move this project to C++. It would be fairly simple to save to xml files there as well. The only issues are HeroForge is pretty big at the moment and I'm not sure how long it would take to transfer all the logic over. Making the project in Qt would allow it to run on most OSs.

**Comment by Heliomance** [comment]
Unfortunately I don't know C++, and I'd rather not just abandon the project to someone else. Also, I hold the extreme user customisability as a fundamental design aim - switching to a compiled format would rather cripple that. I do plan to eventually move to a database format, but for now I'm sticking with Excel.

**Comment by chaoscommencer** [comment]
Well, I'm a C++ software engineer and could help with an alternative version if you'd like. Just don't know how long it would take.

**Comment by torrasque666** [comment]
Just an updated picture with some examples
![image](https://cloud.githubusercontent.com/assets/10926262/9488939/00e06c8c-4ba3-11e5-8f42-69cb95612a75.png)

**Comment by TwistedFaith84** [comment]
You know if the feats page is slow cause of a ton of checkboxes why not just change the checkboxes to 1s and 0s. 1s=Selected,0s=Not Selected. The only prob I can see there is having to change all the formulas to reflect that small (BUT HUGE) change.

**Comment by Heliomance** [comment]
Because that's way less intuitive and easy-to-use.

**Comment by TwistedFaith84** [comment]
@torrasque666 you have a good idea going but that's is a massive amounts of feats and sometimes its good to see the feat and the short descriptive of what it does b4 you select that feat. That way you can remind yourself what that feat does.

**Comment by torrasque666** [comment]
True. I was planning on adding a section on the side to view that, probably in addition to the prerequisites area.

As for the size, I went with 240 as I had a roughly estimated that to be the maximum possible number of feats that could be selected. They can be grouped and minimized before I actually say its done.
